File: WaterCueWindows/Services/NotificationService.cs
using Hardcodet.Wpf.TaskbarNotification;
using System.Windows;

namespace WaterCueWindows.Services;

public class NotificationService
{
    public static readonly NotificationService Shared = new();

    public Action? OnDrinkNow { get; set; }

    private TaskbarIcon? _trayIcon;

    private NotificationService() { }

    public void Initialize(TaskbarIcon icon)
    {
        _trayIcon = icon;
        _trayIcon.TrayBalloonTipClicked += OnBalloonClicked;
    }

    public void ScheduleWarning(int secondsUntilLock)
    {
        var minutes = Math.Max(1, secondsUntilLock / 60);
        ShowBalloon("WaterCue", $"Beba água agora. A tela trava em {minutes} minuto(s).");
    }

    public void CancelWarning()
    {
    }

    public void SendUnlockCelebration()
    {
        ShowBalloon("WaterCue", "Boa! Hidratação validada. Até a próxima.");
    }

    private void ShowBalloon(string title, string message)
    {
        Application.Current?.Dispatcher.Invoke(() =>
        {
            var icon = _trayIcon ?? TryGetTrayIcon();
            icon?.ShowBalloonTip(title, message, BalloonIcon.Info);
        });
    }

    private static TaskbarIcon? TryGetTrayIcon()
        => Application.Current?.Resources["TrayIcon"] as TaskbarIcon;

    private void OnBalloonClicked(object? sender, RoutedEventArgs e)
    {
        OnDrinkNow?.Invoke();
    }
}

File: WaterCueMac/WaterCue/Services/NotificationService.swift
import UserNotifications
import Foundation

@MainActor
final class NotificationService: NSObject, ObservableObject {
    static let shared = NotificationService()

    private let center = UNUserNotificationCenter.current()
    var onDrinkNow: (() -> Void)?

    override init() {
        super.init()
        center.delegate = self
        setupCategories()
    }

    func requestPermission() async -> Bool {
        do {
            return try await center.requestAuthorization(options: [.alert, .sound, .badge])
        } catch {
            return false
        }
    }

    var isAuthorized: Bool {
        get async {
            let settings = await center.notificationSettings()
            return settings.authorizationStatus == .authorized
        }
    }

    private func setupCategories() {
        let drinkNow = UNNotificationAction(
            identifier: "DRINK_NOW",
            title: "Já vou beber",
            options: [.foreground]
        )
        let category = UNNotificationCategory(
            identifier: "PRE_LOCK",
            actions: [drinkNow],
            intentIdentifiers: [],
            options: []
        )
        center.setNotificationCategories([category])
    }

    func scheduleWarning(in seconds: TimeInterval) {
        center.removePendingNotificationRequests(withIdentifiers: ["watercue-warning"])

        let content = UNMutableNotificationContent()
        content.title = "WaterCue"
        content.body = "Beba água agora — tela trava em \(Int(seconds / 60)) minuto(s)!"
        content.sound = .default
        content.categoryIdentifier = "PRE_LOCK"

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: max(1, seconds), repeats: false)
        let request = UNNotificationRequest(identifier: "watercue-warning", content: content, trigger: trigger)
        center.add(request)
    }

    func cancelWarning() {
        center.removePendingNotificationRequests(withIdentifiers: ["watercue-warning"])
    }

    func sendUnlockCelebration() {
        let content = UNMutableNotificationContent()
        content.title = "WaterCue"
        content.body = "Boa! Hidratação validada. Até a próxima."
        content.sound = .default

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 0.1, repeats: false)
        let request = UNNotificationRequest(identifier: "watercue-unlock-\(Date().timeIntervalSince1970)",
                                            content: content, trigger: trigger)
        center.add(request)
    }
}

extension NotificationService: UNUserNotificationCenterDelegate {
    nonisolated func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {
        if response.actionIdentifier == "DRINK_NOW" {
            Task { @MainActor in
                self.onDrinkNow?()
            }
        }
        completionHandler()
    }

    nonisolated func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void
    ) {
        completionHandler([.banner, .sound])
    }
}
